src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JRTArchive.java

```diff
@@ -437,76 +415,83 @@ interface JRTFile {
+        public Entry toEntry() {
+            assert resPath.startsWith("/" + archive.moduleName() + "/");
+            String resName = resPath.substring(archive.moduleName().length() + 2);
+
+            // If the resource has a diff to the packaged modules, use the diff.
+            // Diffs of kind ADDED have been filtered out in collectFiles();
+            if (diff != null) {
+                assert diff.getKind() != ResourceDiff.Kind.ADDED;
+                assert diff.getName().equals(resPath);
+
+                return new Entry(archive, resPath, resName, EntryType.CLASS_OR_RESOURCE) {
+                    @Override
+                    public long size() {
+                        return diff.getResourceBytes().length;
+                    }
+                    @Override
+                    public InputStream stream() {
+                        return new ByteArrayInputStream(diff.getResourceBytes());
+                    }
+                };
+            } else {
+                return new Entry(archive, resPath, resName, EntryType.CLASS_OR_RESOURCE) {
+                    @Override
+                    public long size() {
+                        return archive.imageResources.sizeOf(path());
+                    }
+
+                    @Override
+                    public InputStream stream() {
+                        return archive.imageResources.open(path());
+                    }
+                };
+            }
+        }
+    }
+
+    record JrtOtherFile(
+            JRTArchive archive,
+            String resPath,
+            EntryType resType,
+            String sha,
+            boolean symlink) implements JRTFile {
+
+        // Read from the base JDK image, special casing
+        // symlinks, which have the link target in the
+        // hashOrTarget field.
+        Path targetPath() {
+            return BASE.resolve(symlink ? sha : resPath);
+        }
+
         public Entry toEntry() {
-            return new Entry(archive,
-                             String.format("/%s/%s",
-                                           archive.moduleName(),
-                                           resPath),
-                             resPath,
-                             resType) {
+            assert resType != EntryType.CLASS_OR_RESOURCE;
+
+            return new Entry(
+                    archive,
+                    String.format("/%s/%s", archive.moduleName(), resPath),
+                    resPath,
+                    resType) {
+
                 @Override
                 public long size() {
                     try {
-                        if (resType != EntryType.CLASS_OR_RESOURCE) {
-                            // Read from the base JDK image, special casing
-                            // symlinks, which have the link target in the
-                            // hashOrTarget field
-                            if (symlink) {
-                                return Files.size(BASE.resolve(sha));
```

Comment:
The hashes were present to ensure the integrity of the dependent resources. 
With them removed, what ensures the integrity of the dependencies?